### PR TITLE
Issue 1022

### PR DESCRIPTION
**Pull request by brsnw250**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: etna/models/nn/nbeats/utils.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    Returns
+    -------
+    :
+        Input data as tensor.
+    """
+    return torch.tensor(x, dtype=torch.float32)
+
+
+def prepare_train_batch(
```
Comment (Mr-Geekman) [inline etna/models/nn/nbeats/utils.py:44]: Is this ok that collate function does this logic? I thought that collate function is only responsible for grouping batches into one object.

File: etna/models/nn/nbeats/utils.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    batch_size = len(data)
+
+    history = np.zeros((batch_size, input_size))
+    history_mask = np.zeros((batch_size, input_size))
+    target = np.zeros((batch_size, output_size))
+    target_mask = np.zeros((batch_size, output_size))
+
+    for i, part in enumerate(data):
+        series = part["history"]
+        cut_point = np.random.randint(low=1, high=len(series) - 1, size=1)[0]
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/models/nn/nbeats/utils.py:57]: It also seems strange that this generator has no connection to torch. Its seed also isn't possible to set.

**Review by Mr-Geekman — CHANGES_REQUESTED** [review]

File: etna/models/nn/nbeats/metrics.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+            Binary mask that denotes which points are valid forecasts.
+
+        Returns
+        -------
+        :
+            Metric value.
+        """
+        ape = torch.abs(y_true - y_pred) / torch.abs(y_true)
+
+        ape[ape != ape] = 0.0
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/models/nn/nbeats/metrics.py:69]: You were going to mark it as todo.

[the remark this answers is not in the document]
Comment (brsnw250) [reply etna/models/nn/nbeats/metrics.py:71]: Added comments

File: etna/models/nn/nbeats/nbeats.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+            test_batch_size=test_batch_size,
+            train_dataloader_params=train_dataloader_params,
+            test_dataloader_params=test_dataloader_params,
+            val_dataloader_params=val_dataloader_params,
+            trainer_params=trainer_params,
+            split_params=split_params,
+        )
+
+
+class NBeatsInterpretableModel(NBeatsBaseModel):
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply etna/models/nn/nbeats/nbeats.py:78]: Added for both models

File: etna/models/nn/nbeats/nbeats.py [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+            dictionary with parameters for :py:func:`torch.utils.data.random_split` for train-test splitting
+                * **train_size**: (*float*) value from 0 to 1 - fraction of samples to use for training
+
+                * **generator**: (*Optional[torch.Generator]*) - generator for reproducibile train-test splitting
+
+                * **torch_dataset_size**: (*Optional[int]*) - number of samples in dataset, in case of dataset not implementing ``__len__``
+        """
+        if isinstance(loss, str):
+            try:
+                self.loss = NBeatsLoss[loss].value
```
[the remark this answers is not in the document]
Comment (brsnw250) [reply etna/models/nn/nbeats/nbeats.py:167]: Reworked exceptions

File: etna/models/nn/nbeats/nbeats.py [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+            val_dataloader_params=val_dataloader_params,
+            trainer_params=trainer_params,
+            split_params=split_params,
+        )
+
+
+class NBeatsInterpretableModel(NBeatsBaseModel):
+    """Interpretable N-BEATS model.
+
+    Paper: https://arxiv.org/pdf/1905.10437.pdf
```
Comment (Mr-Geekman) [inline etna/models/nn/nbeats/nbeats.py:81]: This format renders not very good, you can look at [link](https://deploy-preview-1302--etna-docs.netlify.app/api/etna.models.nn.nbeats.nbeats.nbeatsinterpretablemodel?highlight=nbeatsinterpretablemodel#etna.models.nn.nbeats.nbeats.NBeatsInterpretableModel)

File: tests/test_models/test_inference/test_forecast.py [hunk 7]
```diff
@@ -712,6 +724,18 @@ def test_forecast_out_sample_suffix_failed_deep_state(self, model, transforms, e
...
 
+    @pytest.mark.parametrize(
+        "model,transforms",
+        (
+            (NBeatsInterpretableModel(input_size=7, output_size=7, trainer_params=dict(max_epochs=1)), []),
+            (NBeatsGenericModel(input_size=7, output_size=7, trainer_params=dict(max_epochs=1)), []),
+        ),
+    )
+    def test_forecast_out_sample_suffix_failed_nbeats(self, model, transforms, example_tsds):
+        """This test is expected to fail due to windowed view on data in N-BEATS"""
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply tests/test_models/test_inference/test_forecast.py:735]: This question remains.

[the remark this answers is not in the document]
Comment (brsnw250) [reply tests/test_models/test_inference/test_forecast.py:735]: N-BEATS uses a window of training data that ends immediately before the forecast period. When we do a 2-step forecast (filling the gap and use this data to predict the rest of the forecast period) it creates discrepancies when compared to forecasting for the whole period.

File: etna/models/nn/nbeats/nbeats.py (hunk 6 above)
Comment (brsnw250) [reply etna/models/nn/nbeats/nbeats.py:90]: Fixed formatting a bit.

File: tests/test_models/test_inference/test_forecast.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply tests/test_models/test_inference/test_forecast.py:735]: Ok, understood.

File: etna/models/nn/nbeats/utils.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    history = np.zeros((batch_size, input_size))
+    history_mask = np.zeros((batch_size, input_size))
+    target = np.zeros((batch_size, output_size))
+    target_mask = np.zeros((batch_size, output_size))
+
+    for i, part in enumerate(data):
+        series = part["history"]
+
+        if window_sampling_limit is not None:
+            lower_bound = max(1, len(series) - window_sampling_limit - 2)
```
Comment (Mr-Geekman) [inline etna/models/nn/nbeats/utils.py:61]: Why is -2?

**Review by Mr-Geekman — CHANGES_REQUESTED** [review]
One little question.

File: etna/models/nn/nbeats/utils.py (hunk 8 above)
Comment (brsnw250) [reply etna/models/nn/nbeats/utils.py:61]: To ensure at least 1 point, when `window_sampling_limit = 0`

Comment (Mr-Geekman) [reply etna/models/nn/nbeats/utils.py:61]: It can seem a little bit strange to work with `window_sampling_limit=0`. Why not to work with negative values? I think that this value should represent some size, but now it represents it with non-obvious bias.

**Review by Mr-Geekman — APPROVED** [review]

**Review by Mr-Geekman — APPROVED** [review]

**Review by martins0n — APPROVED** [review]
